src/engine/Match_tainting_mode.ml

```diff
@@ -132,6 +132,87 @@ type propagator_match = {
 }
 (** Taint will flow from `from` to `to_` through the axiliary variable `id`. *)
 
+(* This Formula_tbl structure is used to create a "formula cache", which will
+   permit the sharing of matches resulting from common formulas that occur as
+   sources, sinks, sanitizers, or propagators.
+
+   In particular, because of hardcoded propagators, we expect to see lots of
+   sharing from Semgrep Pro Engine.
+*)
+module Formula_tbl = struct
+  include Hashtbl.Make (struct
+    type t = Rule.formula
+
+    let equal = AST_utils.with_structural_equal Rule.equal_formula
+    let hash = Rule.hash_formula
+  end)
+
+  let cached_find_opt formula_cache formula compute_matches_fn =
+    match find_opt formula_cache formula with
+    | None ->
+        (* it should not actually be possible for a formula to
+           not be in the formula table
+
+           just don't cache it I guess
+        *)
+        raise Common.Impossible
```

Comment:
`logger#error` instead and `comptue_matches_fn` anyways, just make it clear this is defensive programming... but we never know, it's always funny when you get those "impossible" exceptions.

Reply:
oops, i originally had it as `compute_matches_fn` anyways, but i changed it to impossible while debugging. good catch.